File: golden-paths/elixir/plug/lib/golden_plug/router.ex
defmodule GoldenPlug.Router do
  @moduledoc "Golden-path contract in one router: /health /ready /hello /metrics."
  use Plug.Router

  @service "golden-elixir-plug"

  plug(:match)
  plug(:dispatch)

  get "/health" do
    send_json(conn, %{status: "ok"})
  end

  get "/ready" do
    send_json(conn, %{status: "ready"})
  end

  get "/hello" do
    send_json(conn, %{service: @service, message: "hello, weyland"})
  end

  # Prometheus text exposition, hand-rolled to stay dependency-free.
  get "/metrics" do
    body =
      "# HELP golden_hello_requests_total Calls to /hello\n" <>
        "# TYPE golden_hello_requests_total counter\n" <>
        "golden_hello_requests_total 0\n"

    conn
    |> put_resp_content_type("text/plain; version=0.0.4")
    |> send_resp(200, body)
  end

  match _ do
    send_resp(conn, 404, "not found")
  end

  defp send_json(conn, data) do
    conn
    |> put_resp_content_type("application/json")
    |> send_resp(200, Jason.encode!(data))
  end
end

File: golden-paths/dart/shelf/lib/router.dart
// Golden path — server-side Dart / shelf (B160).
//
// The idiomatic shelf + shelf_router baseline. Runnable, ephemeral, extendable. Conforms to the
// golden-path contract (docs/design/golden-paths.md): GET /health /ready /metrics /hello.
// Scaffold FROM it: scripts/new-service.sh dart/shelf <your-service>.
library;

import 'dart:convert';

import 'package:shelf/shelf.dart';
import 'package:shelf_router/shelf_router.dart';

const serviceName = 'golden-dart-shelf';

/// Calls to the demo /hello endpoint. Library-global so it survives across router builds.
int helloHits = 0;

Response _json(Map<String, String> body) => Response.ok(
      jsonEncode(body),
      headers: {'content-type': 'application/json'},
    );

/// Renders a minimal-but-valid Prometheus text exposition (format version 0.0.4).
String _renderMetrics() {
  final buf = StringBuffer()
    ..writeln('# HELP golden_hello_requests_total Calls to the demo /hello endpoint')
    ..writeln('# TYPE golden_hello_requests_total counter')
    ..writeln('golden_hello_requests_total $helloHits');
  return buf.toString();
}

/// Builds the contract handler. Returned as a [Handler] so tests drive it directly
/// (no live socket needed) and `bin/server.dart` serves it over shelf_io.
Handler buildRouter() {
  final router = Router();

  router.get('/health', (Request _) => _json({'status': 'ok'}));
  router.get('/ready', (Request _) => _json({'status': 'ready'}));
  router.get('/hello', (Request _) {
    helloHits++;
    return _json({'service': serviceName, 'message': 'hello, weyland'});
  });
  router.get('/metrics', (Request _) => Response.ok(
        _renderMetrics(),
        headers: {'content-type': 'text/plain; version=0.0.4; charset=utf-8'},
      ));

  return router.call;
}
